File: Algorithm/C#/review/Algorithm2/Algorithm2/Containers/PriorityQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;

/// <summary>
/// How to design class:
/// 1\ Design API first.
/// 2\ Decide the data structure you use to represent data.
/// 3\ Method as member of class.
/// </summary>
namespace Algorithm2.Containers
{
    /// <summary>
    /// 1\ Design template class
    /// 2\ General usage by passing a comparator.
    /// 3\ Outline your whole process first and add details after. Abstract your method by private function.
    /// 
    /// complexity:
    /// push/pop: O(lgN)
    /// </summary>
    /// <typeparam name="T"></typeparam>
    public class PriorityQueue<T>
    {
        IList<T> arr;
        Comparison<T> comparator;
        public PriorityQueue(Comparison<T> comparator)
        {
            arr = new List<T>();
            this.comparator = comparator;
        }

        public void Push(T item)
        {
            arr.Add(item);
            Swim(arr.Count() - 1);
        }

        public T Front()
        {
            if(arr.Count() == 0)
            {
                throw new InvalidOperationException();
            }
            return arr[0];
        }

        public void Pop()
        {
            var N = arr.Count();
            if (N == 0)
            {
                throw new InvalidOperationException();
            }
            Swap(0, N - 1);
            arr.RemoveAt(N - 1);
            Sink(0);
        }

        public int Size()
        {
            return arr.Count();
        }
    }
}

File: Algorithm/C++/Algorithm/Algorithm/PriorityQueue.cpp
#include <iostream>
#include <queue>
using namespace std;

template<typename T>
class PriorityQueue {
private:
	vector<T> arr;
	std::function<int(T, T)> comparator;
	void swim(int i);
	void sink(int i);

public:
	PriorityQueue(std::function<int (T, T)> func);
	void push(T item);
	void pop();
	T front();
	int size();
};

template<typename T> PriorityQueue::PriorityQueue(std::function<int(T, T)> func)
{
	comparator = func;
}

template<typename T> void PriorityQueue::push(T item)
{
	arr.push_back(item);
	swim(arr.size() - 1);
}

template<typename T> void PriorityQueue::pop()
{
	if (arr.size() == 0)
	{
		throw exception();
	}
	swap(arr, 0, arr.size() - 1);
	sink(0);
	arr.pop_back();
}

template<typename T> T PriorityQueue::front()
{
	return arr[0];
}

template<typename T> int PriorityQueue::size()
{
	return arr.size();
}

template<typename T> void PriorityQueue::swim(int i)
{
	while (i > 0 && comparator(arr[i], arr[(i - 1) / 2)] > 0)
	{
		swap(arr, i, (i - 1) / 2);
		i = (i - 1) / 2;
	}
}

template<typename T> void PriorityQueue::sink(int i)
{
	while (2 * i + 1 < arr.size())
	{
		int minIdx = comparator(arr[i], arr[2 * i + 1]) > 0 ? i : 2 * i + 1;
		if (2 * i + 2 < arr.size())
		{
			minIdx = comparator(arr[minIdx], arr[2 * i + 2]) > 0 ? minIdx : 2 * i + 2;
		}
		if (minIdx == i)
		{
			break;
		}
		swap(arr, i, minIdx);
		i = minIdx;
	}
}